torch/ao/quantization/dynamic_tracing/auto_trace.py

```diff
@@ -0,0 +1,554 @@
+                    module_stack.append(self)
+                    fqn = module_id_to_fqn.get(id(self), None)
+
+                    if enable_logging:
+                        fqn = module_id_to_fqn.get(id(self), None)
+                        logger.debug(f"\nstarting fqn {fqn}")
+
+                    hook_type = get_module_hook_type(parent_module, cur_module)
+
+                    if hook_type is HookType.OP_HOOKS:
+                        assert parent_module is not None
+                        parent_qstate = parent_module._auto_quant_state
+                        assert isinstance(parent_qstate, AutoQuantizationState)
+                        # before hooks
+                        if not first_call:
+                            parent_qstate.validate_cur_op(cur_module)
+                        args, kwargs = parent_qstate.op_prepare_before_hook(
+                            cur_module, args, kwargs, first_call, qtensor_id,
+                            fqn, cur_module)
+
+                        # original forward
+                        output = orig_module_call(self, *args, **kwargs)
+
+                        # after hooks
+                        # TODO is it correct to call_cur_module twice here?
+                        output = parent_qstate.op_prepare_after_hook(
+                            cur_module, output, args, first_call, qtensor_id,
+                            cur_module)
+                        parent_qstate.mark_cur_op_complete(cur_module)
+
+                    elif hook_type is HookType.MODULE_IO_HOOKS:
+                        # TODO(future PR): add inputs io hook
+
+                        cur_qstate = cur_module._auto_quant_state
+                        cur_qstate.validate_is_at_first_idx()
+
+                        # original forward
+                        output = orig_module_call(self, *args, **kwargs)
+
+                        # after hooks
+                        assert isinstance(cur_qstate, AutoQuantizationState)
+                        output = cur_qstate.outputs_prepare_hook(
+                            output, first_call, qtensor_id)
+                        cur_qstate.validate_is_at_last_seen_idx()
+
+                    elif hook_type is HookType.ARG_DEQUANTS:
+                        output = orig_module_call(self, *args, **kwargs)
+                        # if this fp32 was inplace, make sure to set the output dtype
+                        # back to torch.float
+                        if hasattr(output, '_qtensor_info'):
+                            del output._qtensor_info
+
+                    else:
+                        output = orig_module_call(self, *args, **kwargs)
+
+                    if enable_logging:
+                        fqn = module_id_to_fqn.get(id(self), None)
+                        logger.debug(f"\nending fqn {fqn}")
+
+                    return output
+                finally:
```

Comment:
wondering why do we use `try finally` here, would this swallow all exceptions? wondering if we can use some other implementations like context manager?

Reply:
hmm, not sure I understand the question fully yet, would love to hear more thoughts.  Is there a specific case you are thinking of where a context manager would make more sense?

`finally` would not swallow exceptions, the code in the `finally` clause is run no matter what even if an exception or an early return happened in the `try clause`.